### ISSUE TITLE
Subscribe potentially locks event loop

### ISSUE THREAD

**Issue by rcastill** [bytebeamio/rumqtt#263]
Maybe this is an user problem more than it is a library design issue, but I'll leave the following use case here in the event that it wasn't discussed yet.

If an user decides to subcribe to topics upon confirmed connection, this could potentially lock the event loop if the requests async channel is full. Reproducible example:

```rust
use std::time::Duration;

use rumqttc::{AsyncClient, ConnectReturnCode, Incoming, MqttOptions, QoS};

async fn publisher(mqcli: AsyncClient) {
    loop {
        println!("Publishing payload");
        if let Err(e) = mqcli
            .publish("some/random/topic", QoS::ExactlyOnce, false, "payload")
            .await
        {
            println!("Could not publish message: {}", e);
        }
        tokio::time::sleep(Duration::from_secs(1)).await;
    }
}

#[tokio::main]
async fn main() {
    let mut opts = MqttOptions::new("SomeRandomId", "localhost", 1883);
    let max_packet_size = 1024 * 1024;
    opts.set_max_packet_size(max_packet_size, max_packet_size)
        .set_keep_alive(15)
        .set_inflight(10)
        .set_request_channel_capacity(3)
        .set_clean_session(true);
    let (mqcli, mut mqev) = AsyncClient::new(opts, 10);
    tokio::spawn(publisher(mqcli.clone()));
    loop {
        match mqev.poll().await {
            Ok(rumqttc::Event::Incoming(Incoming::ConnAck(ack))) => {
                if ack.code == ConnectReturnCode::Success {
                    println!("Successfully connected to mqtt broker");
                    if let Err(_) = mqcli.subscribe("some/other/topic", QoS::ExactlyOnce).await {
                        println!("Could not subscribe");
                    }
                } else {
                    println!("Could not connect to mqtt broker: {:?}", ack.code);
                    tokio::time::sleep(Duration::from_secs(5)).await;
                }
            }
            // On disconnection, wait before reconnection
            Ok(rumqttc::Event::Incoming(Incoming::Disconnect)) => {
                println!("Disconnected from mqtt broker");
                tokio::time::sleep(Duration::from_secs(5)).await;
            }
            Ok(_) => {}
            // On error, wait before reconnection
            Err(e) => {
                println!("Error while polling mqtt eventloop: {}", e);
                tokio::time::sleep(Duration::from_secs(5)).await;
            }
        }
    }
}
```

In the example above, the requests channel size is set to 10. Also `mqcli.subscribe()` is called after receiving successful connection ACK. If connection with mqtt broker is lost, then after >10s (10 messages published), the requests channel will be full. Then, when connection is re-established and connection ACK is received, the subscription request will block the main loop as there are already 10 `Publish` requests pending.

I understand that this can be bypassed somehow by not subscribing at that point, but it seems out of place to me. Where should subscriptions be handled otherwise?

**Comment by qwandor** [comment]
This is a problem not just for subscriptions, but for any message that is sent in response to receiving a message from the event loop. You can work around it by making the `AsyncClient` max queue longer, or introducing another unbounded channel somewhere, but I'm not very satisfied with that either.

**Comment by rcastill** [comment]
Maybe there should be some distinction between requests, where some should go to an unbounded channel?

**Comment by jplatte** [comment]
Maybe just spawn a new task for the subscribe request?

**Comment by qwandor** [comment]
Spawning a task to add an item to a channel seems more costly than just making the channel longer (or unbounded).

**Comment by rcastill** [comment]
Plus it may incur in concurrency problems. I don't know if order is guaranteed when spawning multiple tasks (in the case multiple subscriptions are blocked).

**Comment by jplatte** [comment]
> Spawning a task to add an item to a channel seems more costly than just making the channel longer (or unbounded).

In the grand scheme of things, it's still very cheap. Increasing channel size or even making it unbounded creates a new problem as you just accumulate a large amount of publish messages then and your subscribe message takes super long to go through.

**Comment by jplatte** [comment]
> Plus it may incur in concurrency problems. I don't know if order is guaranteed when spawning multiple tasks (in the case multiple subscriptions are blocked).

What do you mean? If multiple re-subscriptions have to happen in order, do them in order in one spawned task.

**Comment by rcastill** [comment]
> In the grand scheme of things, it's still very cheap. Increasing channel size or even making it unbounded creates a new problem as you just accumulate a large amount of publish messages then and your subscribe message takes super long to go through.

I think the unbounded channel would be for subscriptions, not publishes. That's what I meant by differentiating requests.

> What do you mean? If multiple re-subscriptions have to happen in order (for whatever reason), do them in order in one spawned task.

That is what I did actually, and I implemented it using an unbounded channel of all things.

**Comment by tekjar** [comment]
I haven't gone through all the comments but the main issue seems to be from the restriction `don't block the eventloop`.  I've exposed  eventloop directly to the user with the assumption that if I documented it well, it won't be a footgun and it significantly simplifies the design and there is no need to (inefficiently) send incoming events through a channel to the user. 

It also opens up the whole eventloop state to the user enabling customizations like easily retrieving pending messages from the state and saving them to disk before device shutdown. Eventloop being async also allows it to be multiplexed with other tokio async code.

The easiest workaround if you want to reply from the eventloop to some incoming message is  to increase the request queue size. But this can also deadlock when the queue is full. 

Other option is to use a different channel to send incoming data from eventloop to a different thread and publishing from there. But the same rule applies. Don't block the eventloop. Use try_send. If it fails, that means you are not processing the massages fast enough. Disconnect the eventloop or handle it differently

Maybe it's ok to not expose the eventloop  in the next design? Spawn a background thread where the eventloop runs and provide channel like interface?

```
let (tx, rx) = rumqtt::client(options);
```

**Comment by rcastill** [comment]
@tekjar I now spawned a task to handle subscriptions concurrently but I think is not ideal.

> Use try_send.

I thought of that. But there is only `try_publish`, there is no `try_subscribe` in the API.

**Comment by tekjar** [comment]
@rcastill Just send a PR?

**Comment by rcastill** [comment]
OK, that would be a start.

**Comment by tekjar** [comment]
@rcastill  I've implemented try methods to all the methods

**Comment by tekjar** [comment]
But this issue is much deeper than this workaround. It's a tradeoff between flexibility and preventing the foot gun.

I've mentioned this in the documentation [here](https://docs.rs/rumqttc/0.5.0/rumqttc/#important-notes) but who reads documentation :P

**Comment by rcastill** [comment]
I knew that the problem was to block the event loop. Thing is, one does not expect to be able to block it calling "subscribe". A good example is that publish() had a try_publish() counterpart and subscribe() didn't. That made the user think that they work differently.

**Comment by tekjar** [comment]
> A good example is that publish() had a try_publish() counterpart and subscribe() didn't. That made the user think that they work differently.

Oh yeah. That was just me not implementing stuff that I don't need :)

**Comment by rcastill** [comment]
Yes, I understand. I think this issue is now resolved.

TL;DR

Because of design, the user must be careful of blocking the eventloop indefinitely. Also if there is need to respond to an event, it should be done concurrently to avoid filling up the event channel.
